Pyth py3: partial Python 3 port

prechelt opened the pull request:
Hi Brendon,
I need a good RTF reader for my application, which is in Python 3.
I chose pyth and attempted to migrate it using 'modernize' and **'six'**.
It was far from easy (because the bytes-vs-plainstring-vs-unicode issues are non-obvious for code that is handling RTF -- in particular as I know neither much of RTF nor the design of that code), but I believe have I have now gotten rtf15/reader and xhtml/writer to work.

I have extended the tests to complete end-to-end tests using the above pair of converters and they work correctly on Python **3.4.3** (on Windows) and Python **2.7.8** (on Cygwin).
The other converters I have neither debugged nor tested.

Unless verifying that the other parts still work (at least on Python 2) is too difficult for you, I would be most grateful if you could turn this into a new release. I will be using it in my application in any case, whether by loading it from PyPI (nice) or not (less nice).

I updated **README** a bit (could use a little more updating, I think) and have left bits of documentation in various other places as well (the **py3migration directory** and several **new README files** in particular).

Please check it out -- and thank you for creating pyth in the first place!

  Lutz

prechelt committed:
Result of running  python-modernize -w pyth6

modernize automates most of the changes required to make
Python 2 code compatible with Python 3.
The resulting code will rely on one additional package: six

Some more changes will be required for code such as this one
that handes both binary data and text data without always being
explicit in the original Python 2 code which is which.

The file  modernize_output.txt  captures the diff of these changes
as shown by the modernize call.
The file  modernize_output_strippeddown.txt is a subset of that
previous file (as described in the header of itself).

prechelt committed:
I have repaired many of the bytes-vs-str issues in
pyth\plugins\rtf15\reader.py
pyth\plugins\xhtml\writer.py

The former in particular was tricky because most strings have to be handled
as bytestrings -- but not all of them.
See  http://pythonhosted.org/six/

These two now appear to work for ASCII and 8-bit non-ASCII characters in the RTF file,
at least for a simple RTF file.
Complex files and true Unicode remain to be seen.

prechelt committed:
corrected remaining mistakes in rtf15/reader

It should now work correctly.

prechelt committed:
examples/reading/rtf15.py can now generate reference test output

prechelt committed:
established end-to-end tests for Rtf15Reader

They are based on pairs of files with the same basename in directories
tests/rtfs           (inputfiles) and
tests/rtf-as-html    (reference outputfiles)
The unittest test method for each pair is created dynamically
by test_readrtf15.py and will write the actual output to
yet another file with that same basename in directory
tests/currentoutput  (files holding actual outputs).
Those files are deleted only if the test succeeds and so
can be used for analysis if the test fails.

prechelt committed:
some cleanup of rtf15/reader.py

prechelt committed:
documentation update

brendonh commented on pyth/plugins/rtf15/reader.py:
```diff
@@ -5,15 +5,23 @@
 
-_CONTROLCHARS = set(string.ascii_letters + string.digits + "-*")
-_DIGITS = set(string.digits + "-")
+_CONTROLCHARS = b'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-*'
```
Why are these no longer sets?

prechelt replied:
Because we need them to be bytes, not strings.  (This is an important change!)
All we need is the 'in' operator, so a bytestring works just as well. And I could not find a way to create the proper set that would work on Python 2 and 3 alike.

brendonh commented on pyth/plugins/rtf15/reader.py:
```diff
@@ -139,57 +149,57 @@ def parse(self):
-        digits = []
-        current = chars
-        first = True
+        chars = []  # for output 1
```
These comments could do with being either improved or removed. As-is, even I'm confused by them.

prechelt replied:
I spent hours with the various aspects of getControl. While doing so, I became confident wrt the semantics of next and first (so I gave them nicer names nextbyte and isfirst) but never became quite sure what all the possibilities are what digits and chars may mean/contain. This insecurity is reflected in my vague comments.
Yes, better ones are warranted.

brendonh commented on pyth/plugins/rtf15/reader.py:
```diff
@@ -312,13 +328,14 @@ def handle_Para(self, para):
-        elif self.listLevel < prevListLevel:
-            l = self.listStack.pop()
-            self.listStack[-1].append(l)
+            elif self.listLevel < prevListLevel:
```
Should the indent really have changed here? Was this incorrect before? (It may have been!)

prechelt replied:
I never tested item list handling; XHTMLWriter does not appear to do any.
It was just my local impression how the logic ought to work.

brendonh commented on pyth/plugins/rtf15/reader.py:
```diff
@@ -542,25 +559,27 @@ def handle_ansi_escape(self, code):
         # Ignore ~, -, and _, since they are optional crap.
-        if symbol in '\\{}':
-            self.content.append(unicode(symbol))
+        assert len(symbol) == 1
```
What's this asserting against?

prechelt replied:
That was a "partial program understanding notice".
Just take it out if it makes you nervous.
And while you're at it: Take out line 564 as well; I overlooked it when taking out its various cousins.

brendonh commented:
Hi Lutz! Thanks a bunch for working on this -- it's been requested a few times lately. The changes look generally good to me, and less scary than I'd have expected. Please take a look at the minor things I've commented above, and meanwhile I'll try to find time in the next few days to bang on it in Py2.

I've been meaning to make a new release for a while, perhaps even a 1.0, but I should do some work on the docs and other small things first. I have some vacation time from next week which might work for that.

prechelt committed:
encodings/symbol.py: py3-migrated and improved

- performed changes needed to run under Python 3
- fixed one mistake in decodeTable
- added blank to the codes list
- improved the error message
- added two more test cases:
  - msword-symbol.rtf
  - wordpad-symbol.rtf
- neither of them works correctly, see 'Limitations' section
  in top-level README
- therefore, the corresponding 'correct' outputs
  are empty files (to make the automated tests fail)

prechelt commented:
I have added one more commit, for supporting 'Symbol' font.
See the commit message for details.
And again I have forgotten one cleanup: removing  #for (i, c) in enumerate(input):
Sigh...

prechelt committed:
late final touches, setup.py version set to 0.7.dev77

prechelt committed:
completely reorganized the RTF-to-XHTML-based tests

I have thrown out the complex test files (except the interesting zh-cn.rtf)
as well as the overly simple or redundant test files and have introduced
a set of simple tests (not covering too much functionality at once)
and a regular naming of the test files.
The file names in tests/rtfs now consist of two parts:
program-testcontent.rtf
where program is one of
- msword: Microsoft Word from Microsoft Office 2013 on Windows 7
- librewriter: Writer from LibreOffice 4.4 on Windows 7
- wordpad: Microsoft Wordpad on Windows 7
and testcontent is a name describing roughly what functionality
is tested in the file.

The corresponding found-to-be-correct test outputs are in
tests/rtf-as-html/*.html
where these outputs have been truncated to zero bytes
for test cases with errors in their output.

prechelt commented:
Yet another commit, containing much better tests.
See the commit message for the content and execute tests/test_readrtf15.py to see for yourself.
(List handling needs improvement.)

eukreign committed:
uncommited files??

eukreign committed:
python3 support fixes

prechelt committed:
Merge pull request #1 from damoti/pyth-py3

more py3 updates

gilsondev commented:
So, this migrations its ok?

prechelt committed:
Merge branch 'pyth-py3' of https://github.com/prechelt/pyth into pyth-py3
unfortunately, I am surprised by why this merge is necessary.

randlet committed:
fix PlaintextWriter encoding

EpocDotFr commented:
I would love having news about this project.

Would this PR be merged?

robertour committed:
plaintext seems to work now

prechelt committed:
Merge pull request #2 from robertour/pyth-py3

merge robertour PR of change by Petro-Viron: pyth.plugins.plaintext.writer seems to be working now

prechelt commented:
I'll open a new pull request to replace this one in a minute (because I have moved that development from a separate branch to the master branch in my fork).
The new one will contain a consolidated version that I am going to consider to be pyth 0.7.
